src/firmware/app/control/trajectory_planner.c

```diff
@@ -0,0 +1,356 @@
+    ArcLengthParametrization_t arc_length_parameterization,
+    const float arc_segment_length)
+{
+    for (unsigned int i = 0; i < num_elements; i++)
+    {
+        // Get the 't' value corresponding to the current arc length (to be used for
+        // further computing)
+        float t = shared_polynomial2d_getTValueAtArcLengthOrder3(
+            path, i * arc_segment_length, arc_length_parameterization);
+
+        // Get the X and Y position at the 't' value defined by the arc length
+        traj_elements[i].position = shared_polynomial2d_getValueOrder3(path, t);
+    }
+}
+
+void app_trajectory_planner_getMaxAllowableSpeedProfile(
+    float max_allowable_speed_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS],
+    Polynomial2dOrder3_t path, unsigned int num_elements,
+    ArcLengthParametrization_t arc_length_parameterization, float arc_segment_length,
+    const float max_allowable_acceleration)
+{
+    Polynomial2dOrder2_t first_deriv = shared_polynomial2d_differentiateOrder3(path);
+    Polynomial2dOrder1_t second_deriv =
+        shared_polynomial2d_differentiateOrder2(first_deriv);
+
+    for (unsigned int i = 0; i < num_elements; i++)
+    {
+        // Get the 't' value corresponding to the current arc length (to be used for
+        // further computing)
+        float t = shared_polynomial2d_getTValueAtArcLengthOrder3(
+            path, i * arc_segment_length, arc_length_parameterization);
+
+        // Create the polynomial representing path curvature
+        //                                              1
+        //                              ---------------------------------
+        //                                     abs(x'y'' - y'x'')
+        //        radius of curvature =      ----------------------
+        //                                     (x'^2 + y'^2)^(3/2)
+        //
+        const float numerator =
+            fabs(shared_polynomial1d_getValueOrder2(first_deriv.x, t) *
+                     shared_polynomial1d_getValueOrder1(second_deriv.y, t) -
+                 shared_polynomial1d_getValueOrder2(first_deriv.y, t) *
+                     shared_polynomial1d_getValueOrder1(second_deriv.x, t));
+        const float denominator =
+            pow(pow(shared_polynomial1d_getValueOrder2(first_deriv.x, t), 2) +
+                    pow(shared_polynomial1d_getValueOrder2(first_deriv.y, t), 2),
+                3.0 / 2.0);
+
+        const float radius_of_curvature = 1 / (numerator / denominator);
+
+        max_allowable_speed_profile[i] =
+            sqrt(max_allowable_acceleration * radius_of_curvature);
+    }
+}
+
+void app_trajectory_planner_generateForwardsContinuousVelocityProfile(
+    unsigned int num_segments,
+    float velocity_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS],
+    float max_allowable_speed_profile[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS],
+    const float arc_segment_length, const float max_allowable_acceleration,
+    const float max_allowable_speed)
+{
```

Comment:
Wasn't there a paper or reference implementation for this somewhere you used? Would be useful for future people to link it here.

Reply:
Added link to paper in .cpp and .h files